pkg/kubectl/util/hash/hash_test.go

```diff
@@ -32,11 +32,19 @@ func TestConfigMapHash(t *testing.T) {
 		err  string
 	}{
 		// empty map
-		{"empty data", &v1.ConfigMap{Data: map[string]string{}}, "42745tchd9", ""},
+		{"empty data", &v1.ConfigMap{Data: map[string]string{}, BinaryData: map[string][]byte{}}, "42745tchd9", ""},
 		// one key
 		{"one key", &v1.ConfigMap{Data: map[string]string{"one": ""}}, "9g67k2htb6", ""},
 		// three keys (tests sorting order)
 		{"three keys", &v1.ConfigMap{Data: map[string]string{"two": "2", "one": "", "three": "3"}}, "f5h7t85m9b", ""},
+		// empty binary data map
+		{"empty binary data", &v1.ConfigMap{BinaryData: map[string][]byte{}}, "dk855m5d49", ""},
```

Comment:
Why does this have a different hash from "empty data"? Oh... I bet json.Marshal encodes null for Data in this case, and an empty object in the other one?

Reply:
i think so

Reply:
Can you print the serialization to double check?
Maybe we should also consider nilling-out empty maps before hashing to make this consistent, since we don't *semantically* treat nil and empty as different in the API. Making this change might break backwards compatibility of the hash... or maybe not since Data was omitempty anyway?

Reply:
@mtaufen compare line 95 and 101, i have the encoded strings there. the difference between empty and null `Data` is existing behavior. Are we allowed to change that?

Reply:
Right now, `--append-hash` is only available as an option to the `kubectl create configmap` subcommand, and it looks like we always allocate the map there: https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/configmap.go#L131.

We might get away with encoding nil maps as `{}` (treating them the same as empty maps) in our encoding function. We should look at secrets too, since the same problem probably exists in that encoding function.

@liggitt wdyt?

Reply:
@mtaufen ack, i can make the change quickly if it sounds ok to @liggitt

Reply:
@liggitt WDYT? (please see @mtaufen 's comment above)

Reply:
I wouldn't change existing treatment of data, or mess with nilling out map values. The length check on binaryData seems fine to me.